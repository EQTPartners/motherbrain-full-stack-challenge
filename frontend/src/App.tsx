import React, { useEffect, useState } from 'react';
import { CssBaseline } from '@material-ui/core';

import { Map, MenuBar } from './components';
import { getConfig } from './conifg';
import { InvestmentsResponse, InvestorsResponse, MapTypes } from './types';

export const App: React.FC = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [data, setData] = useState<InvestmentsResponse | InvestorsResponse | null>(null);
  const [title, setTitle] = useState('Loading...');
  const [type, setType] = useState(MapTypes.Investors);
  const [isBackArrowVisible, setIsBackArrowVisible] = useState(false);

  const config = getConfig();

  const fetchInvestors = async (): Promise<InvestorsResponse> => {
    const response = await fetch(`${config.api.url}/investors`);
    return response.json();
  };

  const fetchInvestments = async (name: string): Promise<InvestmentsResponse> => {
    const response = await fetch(`${config.api.url}/investors/${name}/investments`);
    return response.json();
  };

  useEffect(() => {
    fetchInvestors()
      .then(data => {
        setData(data);
        setTitle(`Investors - ${data?.investors?.length}`);
      })
      .catch(err => console.error(err))
      .finally(() => setIsLoading(false));
  }, []);

  const handleInvestorClick = async (name: string) => {
    setIsLoading(true);
    setTitle('Loading...');
    setData(null);
    setType(MapTypes.Investments);
    try {
      const data = (await fetchInvestments(name)) as InvestmentsResponse;
      setIsBackArrowVisible(true);
      setTitle(`${name} - ${data.investments.length} investments`);
      setData(data);
    } catch (err) {
      console.error(err);
      setIsBackArrowVisible(true);
      setTitle('Error :( try something else!');
    }
    setIsLoading(false);
  };

  const handleBackButtonClick = async () => {
    setIsLoading(true);
    setTitle('Loading...');
    setData(null);
    setIsBackArrowVisible(false);
    setType(MapTypes.Investors);
    try {
      const data = await fetchInvestors();
      setData(data);
      setTitle(`Investors - ${data?.investors?.length}`);
    } catch (err) {
      console.error(err);
      setTitle('Error :( try something else!');
    }
    setIsLoading(false);
  };

  return (
    <>
      <CssBaseline />
      <MenuBar
        title={title}
        isBackArrowVisible={isBackArrowVisible}
        handleBackButtonClick={handleBackButtonClick}
      />
      <Map type={type} data={data} handleInvestorClick={handleInvestorClick} />
    </>
  );
};
