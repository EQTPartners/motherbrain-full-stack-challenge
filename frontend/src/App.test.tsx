import React from 'react';
import { render } from '@testing-library/react';

import { App } from './App';

describe('App', () => {
  // Rendering the Deck.GL components isn't working because GL only exists in the browser.
  it.skip('should render', () => {
    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
