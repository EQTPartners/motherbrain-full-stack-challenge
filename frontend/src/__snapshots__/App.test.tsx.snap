// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<header
  class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-root-1 MuiPaper-elevation4"
>
  <div
    class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
  >
    <h6
      class="MuiTypography-root makeStyles-title-3 MuiTypography-h6"
    >
      Investors
    </h6>
  </div>
</header>
`;
